File: tests/domains/ads-stats.test.ts
import { describe, expect, it, mock } from "bun:test";
import { collectStats, analyzeStats } from "../../src/domains/ads-stats.js";

describe("ads stats", () => {
  it("collects daily insights and stores as JSON", async () => {
    const graphApi = mock(async () => ({
      data: [{
        campaign_id: "123", campaign_name: "Test",
        impressions: "10000", clicks: "500", spend: "100.50",
        ctr: "0.05", cpm: "10.05", cpc: "0.201",
        date_start: "2026-03-01", date_stop: "2026-03-01",
      }],
    }));

    const result = await collectStats({ graphApi } as any, "act_123", "TOKEN", "2026-03-01", "2026-03-01", "/tmp/fb-stats-test");
    expect(result.campaigns).toBe(1);
    expect(result.dataPoints).toBeGreaterThan(0);
  });

  it("calculates min/max/avg/stddev from stored data", () => {
    const data = [
      { impressions: 100, clicks: 10, spend: 5.0 },
      { impressions: 200, clicks: 20, spend: 10.0 },
      { impressions: 300, clicks: 30, spend: 15.0 },
    ];
    const analysis = analyzeStats(data);
    expect(analysis.impressions.avg).toBe(200);
    expect(analysis.impressions.min).toBe(100);
    expect(analysis.impressions.max).toBe(300);
  });
});

File: internal/facebook/ads_stats_test.go
package facebook

import "testing"

func TestToNumberParsesString(t *testing.T) {
	got := toNumber("123.45")
	if got != 123.45 {
		t.Fatalf("expected 123.45, got %v", got)
	}
}

func TestAnalyzeStatsHandlesStringNumbers(t *testing.T) {
	result := AnalyzeStats([]map[string]any{
		{"impressions": "10", "clicks": "2", "spend": "1.5", "ctr": "0.2", "cpc": "0.75", "cpm": "150"},
		{"impressions": "20", "clicks": "4", "spend": "3.0", "ctr": "0.2", "cpc": "0.75", "cpm": "150"},
	})
	impressions := result["impressions"].(map[string]any)
	if impressions["max"] != 20.0 {
		t.Fatalf("expected max impressions 20, got %#v", impressions["max"])
	}
}
